add `:meander.epsilon/fast` flag

ribelo opened the pull request:
Disable type checks and bound checks and use fastest possible functions
for lookup

ribelo committed:
add `:meander.epsilon/fast` flag

Disable type checks and bound checks and use fastest possible functions
for lookup and nth

noprompt committed:
Generate dissoc calls only when necessary

noprompt commented:
Based on some discussion in the Slack channel, I think we should build on this a little more.

First, I think there are roughly 4 levels of safety (see the table below) that we can derive by considering the combinations of enabling/disabling type checking/native method calling: benign, safe, unsafe, and dangerous. I think what we should do is the following.

- [x] Add the flag `::m/use-native-methods` which by default is true.
- [x] Desugar the flag `::m/unsafe` to `#::m{:use-native-methods false, :no-type-check true}`.
- [x] Desugar the flag `::m/dangerous` to `#::m{:use-native-methods true, :no-type-check true}`.
- [ ] Update `meander.match.ir.epsilon/unsafe?`
- [ ] Add `meander.match.ir.epsilon/dangerous?`

```
+-------------+----------------+----------------------------+-----------+
| type-check? | native-method? | Result                     | Name      |
+-------------+----------------+----------------------------+-----------+
| true        | false          | (f object ,,,)             | Slow      |
+-------------+----------------+----------------------------+-----------+
| true        | true           | (.method ^Type object ,,,) | Safe      |
+-------------+----------------+----------------------------+-----------+
| false       | false          | (f object ,,,)             | Unsafe    |
+-------------+----------------+----------------------------+-----------+
| false       | true           | (.method ^Type object ,,,) | Dangerous |
+-------------+----------------+----------------------------+-----------+
```

noprompt commented:
@ribelo By desugar I meant adding the following functions to the `meander.environment.epsilon` namespace.

```clj
(defn desguar-unsafe [env]
  (if (:meander.epsilon/unsafe env)
    (assoc (dissoc :meander.epsilon/unsafe)
           :meander.epsilon/no-type-check true
           :meander.epsilon/use-native-methods false)
    env))

(defn desguar-dangerous [env]
  (if (:meander.epsilon/dangerous env)
    (assoc (dissoc :meander.epsilon/unsafe)
           :meander.epsilon/no-type-check true
           :meander.epsilon/use-native-methods true)
    env))

(defn desugar [env]
  (-> env
      desugar-unsafe
      desugar-dangerous))
```

Then use the `desugar` function in `m.match/analyze-match-args` and similar functions where we initialize the environment such as [here](https://github.com/noprompt/meander/blob/epsilon/src/meander/match/epsilon.cljc#L1759).

Additionally, I think we should add these predicates to `meander.environment.epsilon`.

```clj
(defn unsafe? [env]
  (and (true? (:meander.epsilon/no-type-check env))
       (false? (:meander.epsilon/use-native-methods env))))

(defn dangerous? [env]
  (and (true? (:meander.epsilon/no-type-check env))
       (true? (:meander.epsilon/use-native-methods env))))
```

We update the predicates in `meander.match.ir.epsilon` to use these.

ribelo committed:
Handle `::m/dangerous` & `::m/unsafe` flag

ribelo commented:
@noprompt 

I removed `unsafe?` and `dangerous?` and added `native-method?` It seems to me that because we have `desugar`, the above have no function. I also used `desugar` in a different place than you suggested, because I wanted to use it as early as possible so that all functions that use `env` already have updated variables.

I also noticed that neither `find` nor `search` use `r.environment/default`, so I took the liberty of adding that.

EDIT:
Fix markdown.

noprompt commented:
> I also noticed that neither find nor search use r.environment/default, so I took the liberty of adding that.

ribelo committed:
Merge branch 'noprompt:epsilon' into epsilon

noprompt committed:
Update README.md

ribelo committed:
use default env in rewrite & rewrites

ribelo committed:
desugar in more places

noprompt committed:
Return to consolidating with; Fix issues with map matching

Closes GH-194

noprompt committed:
Add test for GH-195

noprompt committed:
Ensure that & doesn't get parsed as :amp outside of collections

Fixes GH-199

noprompt committed:
Update README.md

noprompt committed:
Update README.md

markaddleman committed:
clojure tools build based build system

markaddleman committed:
add release dir to gitignore

markaddleman committed:
generate library names that match existing naming convention

markaddleman committed:
even more generate library names that match existing naming convention

noprompt committed:
A few small tweaks to the build

noprompt committed:
Update epsilon.cljc

ribelo committed:
Add example: poor man's malli/spec

noprompt commented on src/meander/environment/epsilon.cljc:
```diff
@@ -35,8 +35,27 @@
    :meander.epsilon/substitute-acyclic-references true
    ;; When set to true disables type checks and bounds checks.
+   :meander.epsilon/use-native-methods true
    :meander.epsilon/unsafe false
```
We can remove these two fields because their meaning is given by the fields they desugar to.

noprompt commented on src/meander/match/ir/epsilon.cljc:
```diff
@@ -34,24 +34,25 @@ compilation decisions."
-  (boolean (:meander.epsilon/unsafe *env*)))
+  (:meander.epsilon/use-native-methods *env*))
+
+(defn fast?
```
We can delete this function.

noprompt commented:
@ribelo From what I can tell this seems good to me. The only thing left would be to delete those fields and the `fast?` predicate and you can hit the merge button. :smile:

ribelo committed:
Remove `fast?` fn

ribelo committed:
Remove unnecessary env keys
